File: apps/hospital_saas/hospital_saas/hospital_saas/report/radiology_orders_report/radiology_orders_report.js
// Copyright (c) 2024, Hospital SAAS and contributors
// For license information, please see license.txt

frappe.query_reports["Radiology Orders Report"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "modality",
            "label": __("Modality"),
            "fieldtype": "Select",
            "options": "\nX-Ray\nCT Scan\nMRI\nUltrasound\nMammography\nFluoroscopy\nNuclear Medicine\nPET Scan\nDEXA Scan\nAngiography"
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nOrdered\nScheduled\nIn Progress\nCompleted\nCancelled"
        },
        {
            "fieldname": "priority",
            "label": __("Priority"),
            "fieldtype": "Select",
            "options": "\nRoutine\nUrgent\nSTAT"
        },
        {
            "fieldname": "examination_type",
            "label": __("Examination Type"),
            "fieldtype": "Link",
            "options": "Radiology Examination Type"
        },
        {
            "fieldname": "practitioner",
            "label": __("Ordering Doctor"),
            "fieldtype": "Link",
            "options": "Healthcare Practitioner"
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname == "priority") {
            if (data.priority == "STAT") {
                value = `<span class="badge" style="background-color: #ef4444; color: white;">${value}</span>`;
            } else if (data.priority == "Urgent") {
                value = `<span class="badge" style="background-color: #f97316; color: white;">${value}</span>`;
            } else {
                value = `<span class="badge" style="background-color: #3b82f6; color: white;">${value}</span>`;
            }
        }

        if (column.fieldname == "status") {
            let color = "#64748b";
            if (data.status == "Completed") color = "#22c55e";
            else if (data.status == "In Progress") color = "#eab308";
            else if (data.status == "Scheduled") color = "#a855f7";
            else if (data.status == "Ordered") color = "#3b82f6";
            else if (data.status == "Cancelled") color = "#ef4444";

            value = `<span class="badge" style="background-color: ${color}; color: white;">${value}</span>`;
        }

        if (column.fieldname == "turnaround_hours" && data.turnaround_hours) {
            let color = "#22c55e"; // Green for good TAT
            if (data.turnaround_hours > 24) color = "#ef4444"; // Red for > 24 hours
            else if (data.turnaround_hours > 12) color = "#f97316"; // Orange for > 12 hours
            else if (data.turnaround_hours > 6) color = "#eab308"; // Yellow for > 6 hours

            value = `<span style="color: ${color}; font-weight: bold;">${value}</span>`;
        }

        return value;
    }
};

File: apps/hospital_saas/hospital_saas/hospital_saas/report/radiology_orders_report/radiology_orders_report.py
# Copyright (c) 2024, Hospital SAAS and contributors
# For license information, please see license.txt

"""
Radiology Orders Report

Comprehensive report for radiology department analytics
"""

import frappe
from frappe import _
from frappe.utils import getdate, add_days, nowdate


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart(filters)
    summary = get_summary(filters)

    return columns, data, None, chart, summary


def get_columns():
    return [
        {
            "fieldname": "name",
            "label": _("Order ID"),
            "fieldtype": "Link",
            "options": "Radiology Order",
            "width": 120
        },
        {
            "fieldname": "order_date",
            "label": _("Order Date"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "patient_name",
            "label": _("Patient"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "examination_type",
            "label": _("Examination"),
            "fieldtype": "Link",
            "options": "Radiology Examination Type",
            "width": 150
        },
        {
            "fieldname": "modality",
            "label": _("Modality"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "priority",
            "label": _("Priority"),
            "fieldtype": "Data",
            "width": 80
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "practitioner_name",
            "label": _("Ordering Doctor"),
            "fieldtype": "Data",
            "width": 140
        },
        {
            "fieldname": "rate",
            "label": _("Rate"),
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "fieldname": "invoiced",
            "label": _("Invoiced"),
            "fieldtype": "Check",
            "width": 80
        },
        {
            "fieldname": "turnaround_hours",
            "label": _("TAT (Hours)"),
            "fieldtype": "Float",
            "precision": 1,
            "width": 100
        }
    ]


def get_data(filters):
    conditions = get_conditions(filters)

    data = frappe.db.sql("""
        SELECT
            ro.name,
            ro.order_date,
            ro.patient,
            ro.patient_name,
            ro.examination_type,
            ro.modality,
            ro.body_part,
            ro.priority,
            ro.status,
            ro.practitioner_name,
            ro.rate,
            ro.invoiced,
            ro.scheduled_date,
            ro.scheduled_time,
            rr.name as result_name,
            rr.examination_date as result_date,
            CASE
                WHEN rr.examination_date IS NOT NULL THEN
                    TIMESTAMPDIFF(HOUR,
                        CONCAT(ro.order_date, ' ', COALESCE(ro.order_time, '00:00:00')),
                        CONCAT(rr.examination_date, ' ', COALESCE(rr.examination_time, '00:00:00'))
                    )
                ELSE NULL
            END as turnaround_hours
        FROM `tabRadiology Order` ro
        LEFT JOIN `tabRadiology Result` rr ON rr.radiology_order = ro.name AND rr.docstatus = 1
        WHERE ro.docstatus = 1
        {conditions}
        ORDER BY ro.order_date DESC, ro.creation DESC
    """.format(conditions=conditions), filters, as_dict=True)

    return data


def get_conditions(filters):
    conditions = []

    if filters.get("from_date"):
        conditions.append("ro.order_date >= %(from_date)s")

    if filters.get("to_date"):
        conditions.append("ro.order_date <= %(to_date)s")

    if filters.get("modality"):
        conditions.append("ro.modality = %(modality)s")

    if filters.get("status"):
        conditions.append("ro.status = %(status)s")

    if filters.get("priority"):
        conditions.append("ro.priority = %(priority)s")

    if filters.get("practitioner"):
        conditions.append("ro.practitioner = %(practitioner)s")

    if filters.get("examination_type"):
        conditions.append("ro.examination_type = %(examination_type)s")

    return "AND " + " AND ".join(conditions) if conditions else ""


def get_chart(filters):
    """Get chart data for modality distribution"""
    conditions = get_conditions(filters)

    modality_data = frappe.db.sql("""
        SELECT
            ro.modality,
            COUNT(*) as count
        FROM `tabRadiology Order` ro
        WHERE ro.docstatus = 1
        {conditions}
        GROUP BY ro.modality
        ORDER BY count DESC
    """.format(conditions=conditions), filters, as_dict=True)

    if not modality_data:
        return None

    return {
        "data": {
            "labels": [d.modality for d in modality_data],
            "datasets": [
                {
                    "name": _("Orders"),
                    "values": [d.count for d in modality_data]
                }
            ]
        },
        "type": "bar",
        "colors": ["#4f46e5"],
        "barOptions": {
            "spaceRatio": 0.4
        }
    }


def get_summary(filters):
    """Get summary statistics"""
    conditions = get_conditions(filters)

    stats = frappe.db.sql("""
        SELECT
            COUNT(*) as total_orders,
            COUNT(CASE WHEN ro.status = 'Completed' THEN 1 END) as completed,
            COUNT(CASE WHEN ro.status IN ('Ordered', 'Scheduled', 'In Progress') THEN 1 END) as pending,
            COUNT(CASE WHEN ro.status = 'Cancelled' THEN 1 END) as cancelled,
            COUNT(CASE WHEN ro.priority = 'STAT' THEN 1 END) as stat_orders,
            COUNT(CASE WHEN ro.priority = 'Urgent' THEN 1 END) as urgent_orders,
            SUM(ro.rate) as total_revenue,
            SUM(CASE WHEN ro.invoiced = 1 THEN ro.rate ELSE 0 END) as invoiced_revenue,
            AVG(
                CASE
                    WHEN rr.examination_date IS NOT NULL THEN
                        TIMESTAMPDIFF(HOUR,
                            CONCAT(ro.order_date, ' ', COALESCE(ro.order_time, '00:00:00')),
                            CONCAT(rr.examination_date, ' ', COALESCE(rr.examination_time, '00:00:00'))
                        )
                    ELSE NULL
                END
            ) as avg_turnaround
        FROM `tabRadiology Order` ro
        LEFT JOIN `tabRadiology Result` rr ON rr.radiology_order = ro.name AND rr.docstatus = 1
        WHERE ro.docstatus = 1
        {conditions}
    """.format(conditions=conditions), filters, as_dict=True)[0]

    return [
        {
            "value": stats.total_orders or 0,
            "indicator": "Blue",
            "label": _("Total Orders"),
            "datatype": "Int"
        },
        {
            "value": stats.completed or 0,
            "indicator": "Green",
            "label": _("Completed"),
            "datatype": "Int"
        },
        {
            "value": stats.pending or 0,
            "indicator": "Orange",
            "label": _("Pending"),
            "datatype": "Int"
        },
        {
            "value": stats.stat_orders or 0,
            "indicator": "Red",
            "label": _("STAT Orders"),
            "datatype": "Int"
        },
        {
            "value": stats.total_revenue or 0,
            "indicator": "Blue",
            "label": _("Total Revenue"),
            "datatype": "Currency"
        },
        {
            "value": round(stats.avg_turnaround or 0, 1),
            "indicator": "Purple",
            "label": _("Avg TAT (Hours)"),
            "datatype": "Float"
        }
    ]
